compiler/src/iree/compiler/Dialect/Flow/Transforms/FormScalarDispatches.cpp

```diff
@@ -0,0 +1,279 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/TopologicalSortUtils.h"
+
+#define DEBUG_TYPE "iree-flow-form-scalar-dispatches"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+/// Pass declaration.
+struct FormScalarDispatchesPass
+    : public FormScalarDispatchesBase<FormScalarDispatchesPass> {
+  using FormScalarDispatchesBase<
+      FormScalarDispatchesPass>::FormScalarDispatchesBase;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect, IREE::Flow::FlowDialect,
+                    linalg::LinalgDialect, tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Return true if type represents a value less than `n` elements.
+static bool isScalarOrTensorOfLinearSizeN(int n, Type type) {
+  if (type.isIntOrIndexOrFloat()) {
+    return true;
+  }
+  if (auto tensorType = type.dyn_cast<RankedTensorType>()) {
+    if (!tensorType.hasStaticShape()) {
+      return false;
+    }
+    return tensorType.getNumElements() <= n;
+  }
+  return false;
+}
+
+/// Return `true` for operations that are to be treated as compute operations.
+static bool isComputeOperation(Operation *op) {
+  MLIRContext *context = op->getContext();
+  if (op->getDialect() == context->getLoadedDialect<linalg::LinalgDialect>()) {
```

Comment:
probably should have arith, complex, etc in here as well

Reply:
Yes... Ill make iterative... pushing too many things here can just create a false dependency between dispatches (sometimes it is just better to clone an operation into a dispatch, like affine.apply operations). `arith` and `complex` do make sense... but just being deliberate.

Reply:
I left this as linalg/tensor for now. We can extend it to arith/math if needed, but I think they are all actually converted to Linalg to funnel it through that. If we move away from that, then we fix this.